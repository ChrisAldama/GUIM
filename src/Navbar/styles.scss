@import '../styles/colors.scss';
@import '../styles/themes.scss';

.whole-page {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: $transparent-gray;
  z-index: 1;
}

.MaxWidthContainer {
  max-width: 1440px;
  padding-right: 20px;
  padding-left: 20px;
  margin-right: auto;
  margin-left: auto;
}
.GUIMNavigation {
  @apply --blue-theme;
  width: 100%;
  & nav {
    display: flex;
    justify-content: space-between;
    flex-direction: row;
    align-items: center;
    & ul {
      list-style-type: none;
      margin: 0;
      & li {
        display: flex;
        align-items: center;
        float: left;
        background: red;
        padding: 8px;
        position: relative;
        cursor: pointer;
        height: 100%;
      }
      & li:hover {
        & a{
          color: var(--dark-gray);
          -webkit-transition: all .5s ease;
          transition: all .5s ease;
        }
      }
    }
  }
  & a{
    text-decoration: none;
    color: black;
    font-size: .9em;
    padding: 0 5px;
    white-space: nowrap;
    font-weight: 300;
  }
}
.navbar-list {
  position: relative;
}
.navbar-list:hover {
  & .navbar-sublist {
    display: block;
  }
  & .navbar-sublist-container {
    display: block;
  }
}
.navbar-sublist-container{
  height: 100%;
  position: absolute;
  top: 28px;
  right: 0px;
  display: none;
}
.navbar-sublist {
  top: 13px;
  right: 0px;
  display:none;
  position: absolute;
  padding: 0;
  cursor: pointer;
  z-index: 1;
  background: red;
  & li {
    width: 100%;
    background-color: $light-gray;
    border-left: 1px solid $chart-gray;
    border-right: 1px solid $chart-gray;
    padding: 15px 8px ;
    & a{
      color: $dark-blue !important;
    }
  }
  & li:first-child{
    border-top: 1px solid $chart-gray;
    -webkit-border-top-left-radius: 4px;
    -webkit-border-top-right-radius: 4px;
    -moz-border-radius-topleft: 4px;
    -moz-border-radius-topright: 4px;
    border-top-left-radius: 4px;
    border-top-right-radius: 4px;
  }
  & li:last-child{
    border-bottom: 1px solid $chart-gray;
    -webkit-border-bottom-left-radius: 4px;
    -webkit-border-bottom-right-radius: 4px;
    -moz-border-radius-bottomleft: 4px;
    -moz-border-radius-bottomright: 4px;
    border-bottom-left-radius: 4px;
    border-bottom-right-radius: 4px;
  }
  & li:hover{
    & a{
      color: $main-blue !important;
    }
  }
}
/*.MobileSublistWidth{
  width: 160px;
}
.SublistWidth{
  width: 180px;
}
.navigation-yals-mobile {
  position: fixed;
  background-color: $color_christmas;
  opacity: 1;
  z-index: 100;
  ul {
    list-style-type: none;
    margin: 0;
    height: 100%;
    width: 100%;
    min-width: 120px;
    li {
      display: flex;
      align-items: center;
      float: left;
      padding: 8px;
      position: relative;
      cursor: pointer;
      height: 100%;
      &:hover {
        a{
          color: $main-blue;
          -webkit-transition: all .5s ease;
          transition: all .5s ease;
        }
      }
    }
  }
  a{
    text-decoration: none;
    color: $white;
    font-size: .9em;
    padding: 0 5px;
    white-space: nowrap;
    font-weight: 300;
  }
}
.navbar-mobile-container {
  width: 100%;
  position: absolute;
  top: 28px;
  right: 0px;
  display: block;
}
.navbar-mobile-sublist {
  top: 10px;
  bottom: 10px;
  right: 0px;
  display: block;
  position: absolute;
  padding: 0;
  cursor: pointer;
  z-index: 1;
  li {
    width: 100%;
    background-color: $color_christmas;
    border-left: 1px solid $chart-gray;
    border-right: 1px solid $chart-gray;
    padding: 15px 8px !important;
    border-bottom: 1px solid $chart-gray;
    &:last-child{
      border-bottom: 1px solid $dark-blue;
    }
    a{
      width: 100%;
      color: $white !important;
    }
  }
}
 */

/* REVERSE */

/*.navbar-reverse-sublist-container{
  height: 100%;
  position: absolute;
  top: 40px;
  right: 0px;
  display: none;
  .triangle-down{
    display: none;
    width: 0;
    height: 0;
    border-left: 12px solid transparent;
    border-right: 12px solid transparent;
    border-top: 9px solid $chart-gray;
    position: absolute;
    bottom: 92px;
    margin-top: 5px;
    z-index: 1;
    &:after{
      content: '';
      width: 0;
      height: 0;
      border-left: 11px solid transparent;
      border-right: 11px solid transparent;
      border-top: 8px solid $light-gray;
      position: absolute;
      bottom: 1px;
      right: -11px;
    }
  }
}
.triangle-left {
  right: 104px;
}
.triangle-right {
  right: 16px;
}
.navbar-reverse-sublist {
  right: 0px;
  display: none;
  position: absolute;
  bottom: 100px;
  padding: 0;
  cursor: pointer;
  z-index: 1;
  li {
    width: 100%;
    background-color: $light-gray;
    border-left: 1px solid $chart-gray;
    border-right: 1px solid $chart-gray;
    padding: 15px 8px ;
    &:first-child{
      border-top: 1px solid $chart-gray;
      -webkit-border-top-left-radius: 4px;
      -webkit-border-top-right-radius: 4px;
      -moz-border-radius-topleft: 4px;
      -moz-border-radius-topright: 4px;
      border-top-left-radius: 4px;
      border-top-right-radius: 4px;
    }
    &:last-child{
      border-bottom: 1px solid $chart-gray;
      -webkit-border-bottom-left-radius: 4px;
      -webkit-border-bottom-right-radius: 4px;
      -moz-border-radius-bottomleft: 4px;
      -moz-border-radius-bottomright: 4px;
      border-bottom-left-radius: 4px;
      border-bottom-right-radius: 4px;
    }
    a{
      color: $dark-blue !important;
    }
    &:hover{
      a{
        color: $main-blue !important;
      }
    }
  }
}
.navbar-reverse-list {
  position: relative;
  &:hover {
    .navbar-reverse-sublist, .triangle-down, .navbar-reverse-sublist-container {
      display: block;
    }
  }
}*/
