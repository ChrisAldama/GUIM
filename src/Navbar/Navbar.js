import React, { Component } from 'react';
import "./styles.scss";

const Link = (props) => {
  return <li className={`${props.className || ""}`}>
    <a href={props.link}
      target={`${props.target || "_self"}`}>
      {props.name}
    </a>
  </li>;
};

const Dropdown = (props) => {
  return (
    <li className="dropdown">
      <a href="#">{props.name}</a>
      <div className="dd-container">
        <ul className="dd-list">
          {props.submodules.map((submod) => <Link className="dd-element" key={submod.name} {...submod}/>)}
        </ul>
      </div>
    </li>
  );
};

class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = { permissions: ["default"] };
    this._buildHeaders();
  }

  _havePermission(permissions, to_validate){
    return (permissions.indexOf(to_validate) >= 0) || to_validate === "default";
  }

  _buildHeaders() {
    const { modules } = this.props;
    if(modules.length < 1){
      return;
    }
    return modules.filter((module) => this._havePermission(this.state.permissions, module.permission))
      .map((module) => {
        if(module.mobile === true) return;
        if ( "submodules" in module ) {
          return <Dropdown key={module.name} {...module}/>;
        }
        return <Link className="navbar-element" key={module.name} {...module}/>;
      });
  }

  render() {
    const props = this.props;
    const headers = (
      <ul className="something">
        { this._buildHeaders() }
      </ul>
    );

    return (
      <div className={`GUIMNavigation ${props.className} ${themes[props.theme] || themes["dark"]}`} >
        <nav onClick={e => e.stopPropagation()}>
          <div className="logo">
            <a href="/">{this.props.logo}</a>
          </div>
          <div>
            {headers}
          </div>
        </nav>
      </div>
    );
  }
}

Navbar.defaultProps = {
  logo: "Logo",
  fixed_top: false,
  className: "",
  theme: "light",
  modules: [
    {
      name: "Home",
      link: "?home=yeap",
      permission: "default",
    },
    {
      name: "About us",
      link: "?about-us=guim",
      permission: "default"
    },
    {
      name: "Contact",
      link: "#",
      permission: "default",
      mobile: false,
      submodules: [
        {
          name: "GUIMDevTeam",
          link: "https://github.com/Intelimetrica/GUIM",
          permission: "default"
        },
        {
          name: "Intelimetrica",
          link: "http://intelimetrica.com",
          permission: "default",
          mobile: true
        }
      ]
    }
  ]
}

const themes = {
  dark: "GUIMNavbarDark",
  light: "GUIMNavbarLight"
}

export default Navbar;
