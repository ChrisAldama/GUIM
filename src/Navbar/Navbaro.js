import React, { Component } from 'react';
import "./styles.scss";
import {
  is_authenticated,
  is_in_sandbox_mode
} from 'common-helpers/user_api_helpers';
import {
  modules_not_logged_in,
  modules_logged_in,
  modules_admin
} from 'common-helpers/modules_navbar_constant';
import { sandbox_warning } from 'common-helpers/sandbox';
import { gaNavbar } from 'common-helpers/google_analytics';
import { isEmpty } from 'lodash';

require('common-styles/general.scss');
require('common-styles/bootstrap.scss');
require('common-styles/navbar_yals.scss');
require('common-styles/sandbox.scss');

const navbarLogo = require('common-images/home_logo.svg');
const listLogo = require('common-images/mis_reportes_ico_list.svg');
const crossImg = require('common-images/Close.svg');
const userImg = require('common-images/Mobile_user.svg');
const notFound = require('common-images/error-404.svg');
const ServerError = require('common-images/error-500.svg');
const mobileElementHeight = '55px';

const Link = (props) => {
  return <li>
    <a href={props.link}
      onClick={() => gaNavbar(`${props.name} Desktop`)}
      target={newWindow(props.name)}>
      {props.name}
    </a>
  </li>;
};

const MobileLink = (props) => {
  return <li style={{height: mobileElementHeight}}>
    <a href={props.link}
      onClick={() => gaNavbar(`${props.name} Mobile`)}
      target={newWindow(props.name)}>
      {props.name}
    </a>
  </li>;
};

const newWindow = (name) => {
  if (name === 'Blog') {
    return '_blank'
  }
  return '_self'
}

const Dropdown = (props) => {
  let submodules = props.submodules;

  return (
    <li className="navbar-list">
      <a href="#">{props.name}</a>
      <div className='navbar-sublist-container Grower'>
        <ul className="navbar-sublist">
          {submodules.map((submod) => <Link key={submod.name} {...submod}/>)}
        </ul>
        <div className='triangle-sublist triangle-right'></div>
      </div>
    </li>
  );
};

class Navbar extends Component {
  constructor(props) {
    super(props);
    this.state = {
      authenticated: false,
      is_admin: false,
      permissions: ['default'],
      fetch_user: false,
      user_name: '',
      mobile: true,
      mobile_list: false
    };
    this._changeSize = this._changeSize.bind(this);
    this._mobileClick = this._mobileClick.bind(this);
    this._buildMobileList = this._buildMobileList.bind(this);
    this._buildGeneralList = this._buildGeneralList.bind(this);
    this._resizeFunction = this._resizeFunction.bind(this);
    this._stopPropagation = this._stopPropagation.bind(this);
    this._closeMobileList = this._closeMobileList.bind(this);
  }

  _closeMobileList() {
    this.setState({mobile_list : false});
  }

  _stopPropagation(e) {
    e.stopPropagation();
  }

  componentWillMount() {
    is_authenticated().then((data) => {
      let {nombre, apellidos, admin} = data.data.user;
      this.setState({
        fetch_user: true,
        authenticated: true,
        is_admin: admin,
        permissions: ['default', 'avaluos'],
        user_name: `${nombre} ${apellidos}`
      });
    }).catch((e) => {
      this.setState({fetch_user: true, permissions: ['default']});
    });
    is_in_sandbox_mode().then(e => {
      if (e.data.bypass) {
        sandbox_warning();
      }
    });
  }

  _havePermission(permissions, to_validate){
    return (permissions.indexOf(to_validate) >= 0) || to_validate === 'default';
  }

  _mobileClick() {
    this.setState({mobile_list : !this.state.mobile_list});
  }

  _buildMobileList(modules) {
    if(isEmpty(modules)){
      return;
    }
    return modules.filter((module) => this._havePermission(this.state.permissions, module.permission))
      .map((module) => {
      if ( module.mobile === false ) {
        return;
      }
      if ( 'image' in module) {
        return (
          <li key={module.name} style={{height: mobileElementHeight}}>
            <a onClick={() => gaNavbar(`Mobile configuraciÃ³n`)} href={module.link}>
              <img src={userImg} height='17px'/>
              <span className='LeftRightMargin'>{module.name}</span>
            </a>
          </li>
        );
      }
      return <MobileLink key={module.name} {...module}/>;
    });
  }

  _buildGeneralList(modules) {
    if(isEmpty(modules)){
      return;
    }
    return modules.filter((module) => this._havePermission(this.state.permissions, module.permission))
      .map((module) => {
      if(module.mobile === true) {
        return;
      }
      if ( 'submodules' in module ) {
        return <Dropdown key={module.name} {...module}/>;
      }
      return <Link key={module.name} {...module}/>;
    });
  }

  _buildHeaders() {
    const {authenticated, user_name, is_admin} = this.state;
    const listDisplay = (this.state.mobile_list) ? 'block' :'none' ;
    const listImage = (this.state.mobile_list) ? crossImg : listLogo;
    let modules = (authenticated) ? modules_logged_in(user_name) : modules_not_logged_in;
    if(is_admin && authenticated){
      modules.splice(modules.length - 2,0,modules_admin);
    }
    if(this.state.mobile === true) {
      return (
          <div>
            <div>
              <a onClick={this._mobileClick}>
                <img className='TopMargin' src={listImage} height='20px'/>
              </a>
              <div className='navbar-mobile-container'>
                <ul className="navbar-mobile-sublist" style={{display: listDisplay}}>
                  {this._buildMobileList(modules)}
                </ul>
              </div>
            </div>
          </div>
      )
    } else {
      return(
        <ul className='FullHeightGrower'>
          {this._buildGeneralList(modules)}
        </ul>
      );
    }
  }

  _changeSize() {
    if(window.matchMedia("(max-width:768px)").matches) {
      this.setState({mobile: true});
    } else {
      this.setState({
        mobile: false,
        mobile_list: false
      });
    }
  }

  _resizeFunction() {
    let resizeTimer
      clearTimeout(resizeTimer);
      resizeTimer = setTimeout(() => {
        this._changeSize();
      }, 250)
    };

  componentDidMount() {
    this._resizeFunction();
    window.addEventListener('resize', this._resizeFunction, false);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this._resizeFunction, false);
  }

  render() {
    let navbar = '';
    const logoHeight = (this.state.mobile) ? '18px' : '26px';
    const navbarHeight = (this.state.mobile) ? '38px' : '51px';

    if (this.state.fetch_user) {
      const headers = this._buildHeaders();
      const backgroundClass = (this.state.mobile_list && this.state.mobile) ? 'whole-page' : '';
      const navbarClass = (this.state.mobile) ? 'navigation-yals-mobile' : 'navigation-yals FullHeightGrower';
      navbar = (
        <div className={backgroundClass} onClick={this._closeMobileList}>
          <div  style={{height: navbarHeight, width:'100%'}}>
            <nav style={{width: '100%'}} className={`${navbarClass} RowDirection MaxWidthContainer SpaceBetween CenterAlignment`}
              onClick={this._stopPropagation}>
              <div className="nav-logo">
                <a href="/"><img src={navbarLogo} height={logoHeight} width='auto'/></a>
              </div>
              <div style={{height: '40px'}}>
                  {headers}
              </div>
            </nav>
          </div>
        </div>
      );
    }
    return (
      <div>
        {navbar}
      </div>
    );
  }
}

export default Navbar;
