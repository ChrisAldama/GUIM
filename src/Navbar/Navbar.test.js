import React from 'react';
import Navbar from './Navbar';
import renderer from 'react-test-renderer';
import { mount, shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('<Navbar />', () => {
  let component;
  const reRender = () => {
    component = renderer.create(
      <Navbar
        logo="Logo"
        fixed_top={false}
        className=""
        theme="light"
        modules={[
          {
            name: "menu1",
            to: "#menu1",
            permission: "default"
          },
          {
            name: "menu2",
            to: "#",
            permission: "default",
            submodules: [{
              name: "submenu1",
              to: "#submenu1",
              permission: "default"
            }]
          }
        ]}
      />
    )};

  it('match dom snapshot', () => {
    reRender();
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});

