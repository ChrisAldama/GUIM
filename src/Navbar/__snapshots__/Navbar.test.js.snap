// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navbar /> match dom snapshot 1`] = `
<div
  className="GUIMNavigation  GUIMNavbarLight"
>
  <nav
    onClick={[Function]}
  >
    <div
      className="logo"
    >
      <a
        href="/"
      >
        Logo
      </a>
    </div>
    <div>
      <ul
        className="something"
      >
        <li
          className="navbar-element"
        >
          <a
            href="#menu1"
            target="_self"
          >
            menu1
          </a>
        </li>
        <li
          className="dropdown"
        >
          <a
            href="#"
          >
            menu2
          </a>
          <div
            className="dd-container"
          >
            <ul
              className="dd-list"
            >
              <li
                className="dd-element"
              >
                <a
                  href="#submenu1"
                  target="_self"
                >
                  submenu1
                </a>
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </div>
  </nav>
</div>
`;
